Magnetic Field for CMZ

JulienDoerner opened the pull request:
This contribution adds the magnetic field for the Galactic Center (GC) of the Milky Way. 

The model is published by Guenduez et al. 2019 (https://doi.org/10.1051/0004-6361/201936081).
This model contributes magnetic fields of several molecular clouds and non-thermal filaments. Furthermore one field for the Radio Arc and another for the intercloud medium are provided. 

We want to use this model for the example of galactic propagation in the 3.2 Paper.

lukasmerten commented on .gitignore:
```diff
@@ -21,9 +21,10 @@ cmake/CMakeFiles/
 *.png
 *.pyc
 
+include/kiss/
```
Please remove include/kiss/ and include/healpix_base/ from the .gitignore file.

Those files are likely tracked because you had the install_prefix set to the CRPropa source path. If you change that to, e.g., usr/local or something not within the git repository there is no need to git-ignore these folders.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+    double phi= std::atan2(y,x);
+    return (sin(phi)+1./eta*cos(phi))*BrAz(r,phi,z,m,B1,eta,R,rr);
+}
+
```
Why are phi and r calculated but not used?
Why is a function used at all if the return value is always 0?

JulienDoerner replied:
This was an old relict. I have deleted the function.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+double CMZField::ByPol(double x, double y, double z, double B1, double a1, double a2, double eta) const{
+    double r= sqrt(pow(x,2.)+pow(y,2.));
+    double phi= std::atan2(y,x);
+    double B2=B1/eta;// B1 is the observed magnetic field stremgth and eta the normalization factor otained from the integrations due to the normalization
```
"B1 is the observed magnetic field stre**n**gth and eta the normalization factor otained from the integrations."

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+}
+
+// Parameter identification:
+// r: radius in zylindrical coordinates
```
**cylindrical **
also in the next occasions

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+// z= z component in zylindircal coordintes
+// phi: phi component in zylindircal coordinates
+// B1: observed magnetic field strength
+// B2: normalization factor according to B1
```
I thought eta is the normalization factor?
Is B2 the normalized field strength or a normalization factor?

JulienDoerner replied:
Yes, B2 is the normalized field strength. I have changed the description.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+// phi_s: orientation angle of the azimuthal pattern if g_s(r1(r,z,a),0,a,Lp,Hp,p), azimuthal angle at infinity of the crest surface if g_s(r,z,a,Lp,Hp,p)                                                                                  
+
+Vector3d CMZField::getMCField(const Vector3d& pos) const {//Field in molecular clouds
+	double pi=M_PI;
```
Why not using M_PI?
That occurs also in some other functions.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,458 @@
+    double a1=70.47*pc;// arcmin-> deg->cm
+    double a2=9.89*pc;// arcmin-> deg-> cm
+    double B1=1.e-3;
+    b.y=ByPol(xx,yy,zz,B1,a1,a2,eta)*1.e-4;//1.e-4 due to the conversion from Gauss to Tesla
```
just use the defined unit gauss instead of 1e-4

lukasmerten reviewed (commented):
I just had a quick view on that PR. Especially I have not cross-checked the the parameters of the individual components with the paper.

I think generally it would be good if you could add some tests for that extensions.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+double CMZField::ByPol(double x, double y, double z, double B1, double a1, double a2, double eta) const{
+    double r= sqrt(pow(x,2.)+pow(y,2.));
+    double phi= std::atan2(y,x);
+    double B2=B1/eta;// B1 is the observed magnetic field strength and eta the normalization factor otained from the integrations due to the normalization
```
Make it a little shorter: "B1 is the observed magnetic field strength and eta the normalization factor."

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+
+// Parameter identification:
+// r: radius in cylindrical coordinates
+// z= z component in cylindircal coordintes
```
coordinates

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// z= z component in cylindircal coordintes
+// phi: phi component in cylindircal coordinates
+// B1: observed magnetic field strength
+// B2: normalization magnetic field strength according to B1
```
Keep it short
"B2: normalized magnetic field strength"

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// phi: phi component in cylindircal coordinates
+// B1: observed magnetic field strength
+// B2: normalization magnetic field strength according to B1
+// m: azimuthal wavenumber m=0 for axissymmetric, m=1 for bisymmetric and m=2 quadrosymmetric, ... field lines
```
m: azimuthal wavenumber (m=0 for axisymmetric, m=1 for bisymmetric, etc field lines)

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// B2: normalization magnetic field strength according to B1
+// m: azimuthal wavenumber m=0 for axissymmetric, m=1 for bisymmetric and m=2 quadrosymmetric, ... field lines
+// a: stricly positive free parameters governing the opening of field lines away from the z-axis
+// L: exponential scale length length of the cloud radius Longitudinal extent
```
scale length 
**l**ongitudinal

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// m: azimuthal wavenumber m=0 for axissymmetric, m=1 for bisymmetric and m=2 quadrosymmetric, ... field lines
+// a: stricly positive free parameters governing the opening of field lines away from the z-axis
+// L: exponential scale length length of the cloud radius Longitudinal extent
+// Lp:scale length of the winding function set to a fraction of the cloud radius  Longitudes of peaks mainly at high |l|
```
"Lp: scale length of the winding function set to a fraction of the cloud radius longitudes of peaks mainly at high |l|"
I do not get this sentence. Please reword or clarify.

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// a: stricly positive free parameters governing the opening of field lines away from the z-axis
+// L: exponential scale length length of the cloud radius Longitudinal extent
+// Lp:scale length of the winding function set to a fraction of the cloud radius  Longitudes of peaks mainly at high |l|
+// H: exponential scale high length of the cloud radius Latitudinal extent
```
scale high length --> scale height
**l**atitudinal

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+// L: exponential scale length length of the cloud radius Longitudinal extent
+// Lp:scale length of the winding function set to a fraction of the cloud radius  Longitudes of peaks mainly at high |l|
+// H: exponential scale high length of the cloud radius Latitudinal extent
+// Hp: exponential scale high  set to a fraction of the cloud vertical size Longitude of peaks mainly at hight |b|
```
scale height

lukasmerten commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,441 @@
+    b.y+=ByPol(xx,yy,zz,B1,a1,a2,eta);
+    b.x+=BxPol(xx,yy,zz,B1,a1,a2,eta);
+    b.z+=BzPol(xx,yy,zz,B1,a1,a2,eta);
+
```
Why is that block commented out?

JulienDoerner replied:
This is an older part of the code. G359.91 is only a possible Nonthermal Filament but might be a SNR. 
I have removed this part at this point. Interessed users could add it for themselve.

lukasmerten reviewed (changes requested):
Overall, it looks almost finished. Most of the comments are just typos.

JulienDoerner commented:
I have changed the typos and shorten the comments.

lukasmerten commented:
@JulienDoerner the PR lools much better now. Is it possible for you to add some more tests of the magnetic field model? E.g. is there a reason to test only the poloidal field and not all components? Sorry for not noticing this earlier but I just took a closer look at the test, when I checked the coverage/coveralls report today.

If that is done, we are probably good to merge.

JulienDoerner commented:
@lukasmerten 
I have changed the structure of the model to use the Vector3 functions instead of doubles for each component. This speeds up the code. 
Additional I have added tests for all components.

TobiasWinchen commented on src/magneticField/CMZField.cpp:
```diff
@@ -0,0 +1,306 @@
+
+    double r1 = 1/(1+a*pos.z*pos.z);
+    double Bs = B1*exp(-r1*r/L);
+    double Br = 2*a*pow(r1,3)*r*pos.z*Bs;
```
Should be pow_integer<3> or r1*r1*r1 for performance reasons, see: https://github.com/CRPropa/CRPropa3/blob/3ca1ba03729b1e8735ab1dba87e22a57d1c2008e/include/crpropa/Common.h#L55

JulienDoerner replied:
yes I have changed it.

lukasmerten replied:
The corresponding header file (crpropa/Common.h) was not included, which led to failing builds. Should be fixed now.

TobiasWinchen commented on src/massDistribution/Ferriere.cpp:
```diff
@@ -222,12 +220,11 @@ double Ferriere::getNucleonDensity(const Vector3d &position) const{
-	// check if any density is activ and give warning if not
-	if(isforHI||isforHII||isforH2 == false){
+	// check if all densities is active and raise warning if not
+	if((isforHI & isforHII & isforH2) == false){
```
You probably mean &&  instead of &

JulienDoerner replied:
this is the way like it is introduced  #336 by @lukasmerten. 
I have just copied the code to resolve a merge conflict.

TobiasWinchen replied:
& is a bitwise and while && is logical and. Bitwise and is not what is wanted here. For boolleans the result is the same, but it is still confusing.

JulienDoerner replied:
okay. I have changed it at this part, but I think this done in some other contexts to.

lukasmerten commented:
I think everything is fixed now. We can merge, when the checks are completed.
